File: 0589-n-ary-tree-preorder-traversal/iterative-depth-first-search/python/solution.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        result = []
        stack = [root]
        while stack:
            node = stack.pop()
            if node:
                result.append(node.val)
                # Push children in reverse order so leftmost child is processed first
                for child in node.children[::-1]:
                    stack.append(child)
        return result

File: 0589-n-ary-tree-preorder-traversal/iterative-depth-first-search/cpp/solution.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution {
public:
    vector<int> preorder(Node* root) {
        vector<int> result;
        std::stack<Node*> stk;
        stk.push(root);

        while (!stk.empty()) {
            Node* node = stk.top();
            stk.pop();
            if (node) {
                result.push_back(node->val);
                // Push children in reverse order so that leftmost child is processed first
                for (auto it = node->children.rbegin(); it != node->children.rend(); it++) {
                    stk.push(*it);
                }
            }
        }
        return result;
    }
};
